src/utils/aws.ts

```diff
@@ -1,43 +1,59 @@
 import { ApiGatewayManagementApi } from 'aws-sdk';
-import { APIGatewayEventRequestContext } from 'aws-lambda';
 import {
   ConnectionAckMessage,
   NextMessage,
   CompleteMessage,
   ErrorMessage,
 } from 'graphql-ws';
+import { ServerClosure } from '../types';
 
-export const sendMessage = (
-  a: {
+export const sendMessage = async (
+  c : ServerClosure,
+  {
+    connectionId: ConnectionId,
+    domainName,
+    stage,
+    message,
+  }: {
     message:
       | ConnectionAckMessage
       | NextMessage
       | CompleteMessage
       | ErrorMessage;
-  } & Pick<
-    APIGatewayEventRequestContext,
-    'connectionId' | 'domainName' | 'stage'
-  >
-) =>
-  new ApiGatewayManagementApi({
+    connectionId: string;
+    domainName: string;
+    stage: string;
+  }
+): Promise<void> => {
+  const api = c.apiGatewayManagementApi ?? new ApiGatewayManagementApi({
     apiVersion: 'latest',
-    endpoint: `${a.domainName}/${a.stage}`,
+    endpoint: `${domainName}/${stage}`,
   })
-    .postToConnection({
-      ConnectionId: a.connectionId!,
-      Data: JSON.stringify(a.message),
-    })
-    .promise();
 
-export const deleteConnection = (
-  a: Pick<
-    APIGatewayEventRequestContext,
-    'connectionId' | 'domainName' | 'stage'
-  >
-) =>
-  new ApiGatewayManagementApi({
+  await api.postToConnection({
+    ConnectionId,
+    Data: JSON.stringify(message),
+  })
+  .promise();
+}
+
+export const deleteConnection = async (
+  c : ServerClosure,
+  {
+    connectionId: ConnectionId,
+    domainName,
+    stage
+  }: {
+    connectionId: string;
+    domainName: string;
+    stage: string;
+  }
+): Promise<void> => {
+  const api = c.apiGatewayManagementApi ?? new ApiGatewayManagementApi({
     apiVersion: 'latest',
-    endpoint: `${a.domainName}/${a.stage}`,
+    endpoint: `${domainName}/${stage}`,
   })
-    .deleteConnection({ ConnectionId: a.connectionId! })
+
+  await api.deleteConnection({ ConnectionId })
     .promise();
+}
```

Comment:
This looks good! Just a few small suggestions that'll help me on the maintenance front

To differentiate between what is a closure and what are arguments related to execution, currying would be ideal.

```tsx
export const deleteConnection = (c: ServerClosure) => async (a: ...)
```

Also, can we stick to picking a subset of properties rather than declaring a new type. The end result would look something like

```tsx
export const deleteConnection =
  (c: ServerClosure) =>
  async (
    a: Pick<
      APIGatewayWebsocketEvent['requestContext'],
      "connectionId" | "domainName" | "stage"
    >
  ) => {
    const api =
      c.apiGatewayManagementApi ??
      new ApiGatewayManagementApi({
        apiVersion: "latest",
        endpoint: `${a.domainName}/${a.stage}`,
      });

    await api.deleteConnection({ ConnectionId: a.connectionId }).promise();
  };
```

Reply:
Appreciate the explanation! Mostly for maintenance reasons, I think using the existing types make sense for now - fewer new types is less work for me.

RE closure forwarding, if it helps, you could think of it this way:

```tsx
const makeIncrementor = () => {
  const state = { value: 0 };

  // Increment takes a single argument
  const increment = (amount) => state.value + amount
  
  return  { increment }
}

// If we choose to pull "increment" out, we need to pass the closure
// The signature of the function doesn't change
const increment = (closure) => (arg) => closure.value + amount

const makeIncrementor = () => {
    const state = { value: 0 };
    return  { increment: increment(state) }
}
```